From 8e47971c02f301631269e346f5810640833d532b Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Sun, 29 Mar 2015 16:55:47 +0800
Subject: bcm-xlp: update the driver of smsc911x

The interrupt entry of smsc911x is 13 in the IRT and the related IRQ
is 40 in XLP208.

Updating the function of read/write register otherwise the
function 'smsc911x_phy_check_loopbackpkt' would be failed.

Signed-off-by: Jack Tan <jack.tan@windriver.com>
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/include/asm/netlogic/xlp-hal/xlp.h b/arch/mips/include/asm/netlogic/xlp-hal/xlp.h
index 9afd135..8fa27af 100644
--- a/arch/mips/include/asm/netlogic/xlp-hal/xlp.h
+++ b/arch/mips/include/asm/netlogic/xlp-hal/xlp.h
@@ -63,6 +63,9 @@
 #define PIC_NAND_IRQ			37
 #define PIC_SATA_IRQ			38
 #define PIC_GPIO_IRQ			39
+#define PIC_GPIO_1_IRQ                  40
+#define PIC_GPIO_2_IRQ                  41
+#define PIC_GPIO_3_IRQ                  42
 
 #define PIC_PCIE_LINK_MSI_IRQ_BASE	44	/* 44 - 47 MSI IRQ */
 #define PIC_PCIE_LINK_MSI_IRQ(i)	(44 + (i))
diff --git a/arch/mips/include/asm/netlogic/xlp_smsc.h b/arch/mips/include/asm/netlogic/xlp_smsc.h
new file mode 100644
index 0000000..0b6f6d0
--- /dev/null
+++ b/arch/mips/include/asm/netlogic/xlp_smsc.h
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2003-2012 Broadcom Corporation
+ * All Rights Reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the Broadcom
+ * license below:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY BROADCOM ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL BROADCOM OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __XLP_SMSC_H__
+#define __XLP_SMSC_H__
+
+void xlp_smsc_gpio_setup(void);
+void xlp_smsc_gpio_reg_ack(int irq);
+
+#endif
diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 56bb5e3..5af03c7 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -70,7 +70,7 @@
 #include "brcm_xlp_cpu.h"
 
 /* These addresses are computed by the nlm_hal_init() */
-unsigned long xlp_io_base;
+unsigned long xlp_io_base = KSEG1 + 0x18000000;
 unsigned long xlp_fmn_base[NLM_MAX_NODES];
 unsigned long xlp_nae_base[NLM_MAX_NODES];
 unsigned long xlp_sae_base[NLM_MAX_NODES];
diff --git a/arch/mips/netlogic/xlp/Makefile b/arch/mips/netlogic/xlp/Makefile
index dc17346..d8f8656 100644
--- a/arch/mips/netlogic/xlp/Makefile
+++ b/arch/mips/netlogic/xlp/Makefile
@@ -1,3 +1,6 @@
+EXTRA_CFLAGS := -DNLM_HAL_LINUX_KERNEL -DNLM_CORTINA_SUPPORT -Wno-maybe-uninitialized
+EXTRA_CFLAGS += -Iarch/mips/netlogic/lib/syslib/include
+
 obj-y				+= setup.o nlm_hal.o dt.o c-xlp.o
 obj-y				+= cpld.o
 obj-y				+= pgwalker.o
@@ -13,3 +16,4 @@ obj-$(CONFIG_NLM_ENABLE_COP2)	+= cop2.o
 obj-$(CONFIG_KGDB)		+= nmi.o
 
 obj-y				+= pci-hot-reset.o
+obj-$(CONFIG_SMSC911X)		+= smsc-init.o
diff --git a/arch/mips/netlogic/xlp/nlm_hal.c b/arch/mips/netlogic/xlp/nlm_hal.c
index 8deb579..a84f563 100644
--- a/arch/mips/netlogic/xlp/nlm_hal.c
+++ b/arch/mips/netlogic/xlp/nlm_hal.c
@@ -133,6 +133,9 @@ static int xlp_irq_to_irt(int irq)
 		devoff = XLP_IO_SATA_OFFSET(0);
 		break;
 	case PIC_GPIO_IRQ:
+	case PIC_GPIO_1_IRQ:
+	case PIC_GPIO_2_IRQ:
+	case PIC_GPIO_3_IRQ:
 		devoff = XLP_IO_GPIO_OFFSET(0);
 		break;
 	case PIC_NAND_IRQ:
@@ -196,6 +199,12 @@ static int xlp_irq_to_irt(int irq)
 				irt = irt + 2; break;
 			case PIC_I2C_3_IRQ:
 				irt = irt + 3; break;
+			case PIC_GPIO_1_IRQ:
+				irt = irt + 1; break;
+			case PIC_GPIO_2_IRQ:
+				irt = irt + 2; break;
+			case PIC_GPIO_3_IRQ:
+				irt = irt + 3; break;
 			}
 		}
 	} else if (irq >= PIC_PCIE_LINK_LEGACY_IRQ(0) &&
diff --git a/arch/mips/netlogic/xlp/smsc-init.c b/arch/mips/netlogic/xlp/smsc-init.c
new file mode 100644
index 0000000..54d2669
--- /dev/null
+++ b/arch/mips/netlogic/xlp/smsc-init.c
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2003-2012 Broadcom Corporation
+ * All Rights Reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the Broadcom
+ * license below:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY BROADCOM ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL BROADCOM OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <asm/netlogic/xlp_smsc.h>
+#include <asm/netlogic/common.h>
+
+#include "nlm_hal_xlp_dev.h"
+#include "nlm_hal.h"
+
+static __inline__ int32_t gpio_reg_read(int node, int regidx)
+{
+	volatile uint64_t mmio;
+	mmio = nlm_hal_get_dev_base(node, 0, XLP_PCIE_GIO_DEV, XLP_GIO_GPIO_FUNC);
+	return nlm_hal_read_32bit_reg(mmio, regidx);
+}
+
+static __inline__ void gpio_reg_write(int node, int regidx, int32_t val)
+{
+	volatile uint64_t mmio;
+	mmio = nlm_hal_get_dev_base(node, 0, XLP_PCIE_GIO_DEV, XLP_GIO_GPIO_FUNC);
+	nlm_hal_write_32bit_reg(mmio, regidx, val);
+}
+
+void xlp_smsc_gpio_setup(void)
+{
+	gpio_reg_write(0, XLP_GPIO_INTEN11,
+		gpio_reg_read(0, XLP_GPIO_INTEN11) | 0x80);
+	gpio_reg_write(0, XLP_GPIO_INT_POLAR1,
+		gpio_reg_read(0, XLP_GPIO_INT_POLAR1) | 0x80);
+	gpio_reg_write(0, XLP_GPIO_INT_TYPE1,
+		gpio_reg_read(0, XLP_GPIO_INT_TYPE1) & ~(0x80));
+
+	/* clear all interrupts */
+	gpio_reg_write(0, XLP_GPIO_INT_STAT1, 0xffffffff);
+}
+
+void xlp_smsc_gpio_reg_ack(int irq)
+{
+	gpio_reg_write(0, XLP_GPIO_INT_STAT1,
+		gpio_reg_read(0, XLP_GPIO_INT_STAT1) & 0x80);
+}
diff --git a/drivers/net/ethernet/smsc/smsc911x.c b/drivers/net/ethernet/smsc/smsc911x.c
index a3413a5..5ff0c09 100644
--- a/drivers/net/ethernet/smsc/smsc911x.c
+++ b/drivers/net/ethernet/smsc/smsc911x.c
@@ -61,6 +61,7 @@
 #include <linux/of_gpio.h>
 #include <linux/of_net.h>
 #include "smsc911x.h"
+#include <asm/netlogic/xlp_smsc.h>
 
 #define SMSC_CHIPNAME		"smsc911x"
 #define SMSC_MDIONAME		"smsc911x-mdio"
@@ -1879,6 +1880,8 @@ static irqreturn_t smsc911x_irqhandler(int irq, void *dev_id)
 	int serviced = IRQ_NONE;
 	u32 temp;
 
+	xlp_smsc_gpio_reg_ack(irq);
+
 	if (unlikely(intsts & inten & INT_STS_SW_INT_)) {
 		temp = smsc911x_reg_read(pdata, INT_EN);
 		temp &= (~INT_EN_SW_INT_EN_);
@@ -2582,6 +2585,8 @@ static int smsc911x_drv_probe(struct platform_device *pdev)
 	/* Ensure interrupts are globally disabled before connecting ISR */
 	smsc911x_disable_irq_chip(dev);
 
+	xlp_smsc_gpio_setup();
+
 	retval = request_irq(dev->irq, smsc911x_irqhandler,
 			     irq_flags | IRQF_SHARED, dev->name, dev);
 	if (retval) {
-- 
1.7.1

